lib/test.js

```diff
@@ -1,232 +1,370 @@
-          passed: this.passed.join('\n  '),
-          message,
-          stack: new Error().stack
-        });
+  test(message, func, options = {}) {
+    options.run = false;
+    const test = new Test(message, func, options);
+    if (options.parallel) {
+      this._runSubtest(test);
+    } else {
+      this.subtestQueue.push(test);
+      if (this.subtestQueue.length === 1) {
+        this._runSubtest(test);
       }
     }
-    this.done = true;
-    if (this.events.done) this.events.done();
-    this.events.done = null;
-  };
-
-  Test.prototype.throws = function(fn, err, message) {
-    let errored = false;
-    let msg = '';
+  }
+
+  testAsync(message, func, options = {}) {
+    Object.assign(options, { async: true });
+    this.test(message, func, options);
+  }
+
+  testSync(message, func, options = {}) {
+    Object.assign(options, { async: false });
+    this.test(message, func, options);
+  }
+
+  _check(type, testFunc, actual, expected, message) {
+    const success = testFunc(actual, expected);
+    this.results.push({
+      type,
+      success,
+      stack: new Error().stack,
+      expected, actual, message,
+    });
+    this._passCheck();
+  }
+
+  same(actual, expected, message) {
+    this._check('equal', compare.equal, actual, expected, message);
+  }
+
+  notEqual(actual, expected, message) {
+    this._check(
+      'notEqual',
+      (actual, expected) => !compare.equal(actual, expected),
+      actual,
+      expected,
+      message
+    );
+  }
+
+  strictSame(actual, expected, message) {
+    this._check('strictEqual', compare.strictEqual, actual, expected, message);
+  }
+
+  assert(value, message) {
+    this._check('assert', v => v, !!value, true, message);
+  }
+
+  assertNot(value, message) {
+    this._check('assertNot', v => v, !value, false, message);
+  }
+
+  fail(message) {
+    // TODO(lundibundi): make fail stop the execution of the
+    // whole test(including any subtests)
+    this._check('fail', () => false, null, null, message);
+    this.end();
+  }
+
+  error(err, message) {
+    this._check(
+      'error',
+      (err) => !(err instanceof Error),
+      err,
+      null,
+      message
+    );
+  }
+
+  type(obj, type, message) {
+    this._check(
+      'type',
+      ([domain, category], type) => domain === type || category === type,
+      [typeof obj, obj.constructor.name],
+      type,
+      message
+    );
+  }
+
+  throws(fn, expected, message) {
+    let actual = null;
     try {
       fn();
     } catch (e) {
-      msg = e.message;
-      errored = err.message === msg;
-    }
-    if (!errored) {
-      this.success = false;
-      metatests.fail({
-        caption: this.caption,
-        type: 'throws',
-        subtests: this.subtests,
-        actual: msg, expected: err.message, message,
-        stack: new Error().stack
-      });
+      actual = e;
     }
-  };
 
-  Test.prototype.doesNotThrow = function(fn, message) {
+    const success =
+      actual &&
```

Comment:
I'd prefer that `expected` error is optional, if `expected` is not provided that should pass if `fn()` invocation thew anything.